### ISSUE TITLE
NuGet packaging somehow using versions prior to build

### ISSUE THREAD

**Issue by mwpowellhtx** [NuGet/Home#7290]
I have a reference to [BumpAssemblyVersions](https://www.nuget.org/packages/BumpAssemblyVersions/) that is happening ``BeforeBuild`` targets. However, NuGet is "successfully creating package" and somehow using the value of my versions before that. IMO this is incorrect and should be using the value after build, don't you think? I'm not sure how that is getting stuck, and/or how I should be replacing that in memory in order for ``NuGet`` to draw the correct value. My custom task is running and effectively re-writing the project file, but this does not seem to be sufficient for ``NuGet`` to do the right thing with what I'm expecting. I thought perhaps this was a missing dependency on something like ``GenerateAssemblyInfo``, but when I do this, whatever watchers are sitting atop the file are aggressively resolving ``BumpAssemblyVersions``.

Here's the complete build log:

```
1>------ Rebuild All started: Project: Kingdom.Collections.ImmutableBitArray, Configuration: Release Any CPU ------
1>G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\obj\Kingdom.Collections.ImmutableBitArray.csproj.nuget.g.targets(8,5): warning MSB4011: "G:\Dev\NuGet\global\packages\bumpassemblyversions\1.1.0\build\BumpAssemblyVersions.targets" cannot be imported again. It was already imported at "G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\Kingdom.Collections.ImmutableBitArray.csproj (29,3)". This is most likely a build authoring error. This subsequent import will be ignored.
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'AssemblyVersion' from '1.8.125.20605' to '1.8.126.20622'
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'FileVersion' from '1.8.125.20605' to '1.8.126.20622'
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'PackageVersion' from '1.8.125.20605' to '1.8.126.20622'
1>Kingdom.Collections.ImmutableBitArray -> G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\bin\Release\netstandard2.0\Kingdom.Collections.ImmutableBitArray.dll
1>G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\obj\Kingdom.Collections.ImmutableBitArray.csproj.nuget.g.targets(8,5): warning MSB4011: "G:\Dev\NuGet\global\packages\bumpassemblyversions\1.1.0\build\BumpAssemblyVersions.targets" cannot be imported again. It was already imported at "G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\Kingdom.Collections.ImmutableBitArray.csproj (29,3)". This is most likely a build authoring error. This subsequent import will be ignored.
1>G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\obj\Kingdom.Collections.ImmutableBitArray.csproj.nuget.g.targets(8,5): warning MSB4011: "G:\Dev\NuGet\global\packages\bumpassemblyversions\1.1.0\build\BumpAssemblyVersions.targets" cannot be imported again. It was already imported at "G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\Kingdom.Collections.ImmutableBitArray.csproj (29,3)". This is most likely a build authoring error. This subsequent import will be ignored.
1>Successfully created package 'G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\bin\Release\Kingdom.Collections.ImmutableBitArray.1.8.125.20605.nupkg'.
1>Done building project "Kingdom.Collections.ImmutableBitArray.csproj".
========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========
```

But here are the key bits that I'm concerned about at the moment:

```
...
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'AssemblyVersion' from '1.8.125.20605' to '1.8.126.20622'
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'FileVersion' from '1.8.125.20605' to '1.8.126.20622'
1>'Kingdom.Collections.ImmutableBitArray.csproj': Bumped 'PackageVersion' from '1.8.125.20605' to '1.8.126.20622'
...
1>Successfully created package 'G:\Source\Kingdom Software\Kingdom.Collections\Migrate\src\Kingdom.Collections.ImmutableBitArray\bin\Release\Kingdom.Collections.ImmutableBitArray.1.8.125.20605.nupkg'.
...
```

**Comment by mwpowellhtx** [comment]
I think I see what's going on. I do need to rewrite the project file as I was doing. However, I also need to be more integrated from a *targets* perspective and connect output from my custom task in order for the build process to consume the correct versions at the correct moment. I'll keep this open in the event there is anything more substantive that comes up, but I think that's the approach I need to adopt.

**Comment by rrelyea** [comment]
@mwpowellhtx 

- what type of project are you packing?
- What technique/tool are you using to build the package?

**Comment by mwpowellhtx** [comment]
Just to reiterate, if I try to connect the dots any sooner than ``GetAssemblyVersion``, I am finding that my task is continually invoked, and I get continually bumping versions. Really, not what I am after whatsoever. So... Packaging really needs to depend on build completion, which can depend on assembly versions, assembly information being generated, that's all well and good. But leave room for end user custom tasks doing their work to bump versions.

**Comment by mwpowellhtx** [comment]
@rrelyea It's nothing fancy; I use the IDE to build/rebuild, etc. I am introducing a custom task to bump the version(s), whether they be assembly attribute variety or CSPROJ based.

**Comment by rohit21agrawal** [comment]
@mwpowellhtx can i see in code/targets what you are trying to do? you need  to be setting the Version or PackageVersion property, and as long as that is set before the ```GenerateNuspec``` task of Pack, which only runs after build, you will get the updated version for your package.

**Comment by mwpowellhtx** [comment]
@rohit21agrawal I'm confused then; when I am "bumping" my versions, it would seem that not all versions are created equal, depending on when generating assembly info sees them, versus packaging, etc. Does that sound accurate?

**Comment by rohit21agrawal** [comment]
i am not sure I understand your question/concern. could you please elaborate in a little more detail?

**Comment by mwpowellhtx** [comment]
I think I've got a hypothesis what's going on. My custom task is re-writing the project with bumped versions, and I was half-expecting those to be the versions used throughout. However, MSBuild is more likely running with the project file as-is prior to my custom task working on it; so any versions that NuGet sees, i.e. during packaging, are the previous ones. I was able to expose my versions via Custom Task Output properties and subsequently connect those with the proper *CSPROJ* ``<Version/>``, ``<PackageVersion/>``, etc, and, at least so far, this seems far better behaved.
